controllers/azuremachine_controller.go

```diff
@@ -0,0 +1,226 @@
+
+	// Create the scope.
+	machineScope, err := capzscope.NewMachineScope(capzscope.MachineScopeParams{
+		Client:       r.Client,
+		Machine:      machine,
+		AzureMachine: azureMachine,
+	})
+	if err != nil {
+		log.Error(err, "failed to create machine scope")
+		return reconcile.Result{}, microerror.Mask(err)
+	}
+
+	defer func() {
+		if err := machineScope.Close(ctx); err != nil && reterr == nil {
+			reterr = microerror.Mask(err)
+		}
+	}()
+
+	// Create the cluster scope
+	// needed to get cluster credentials to act on clusters domain
+	clusterScope, err := capzscope.NewClusterScope(ctx, capzscope.ClusterScopeParams{
+		Client:       r.Client,
+		Cluster:      cluster,
+		AzureCluster: azureCluster,
+	})
+	if err != nil {
+		log.Error(err, "failed to create cluster scope")
+		return reconcile.Result{}, microerror.Mask(err)
+	}
+
+	defer func() {
+		if err := clusterScope.Close(ctx); err != nil && reterr == nil {
+			reterr = microerror.Mask(err)
+		}
+	}()
+
+	// only continue reconcilation if the bastion label is set
+	azureMachineLabels := machineScope.AzureMachine.GetLabels()
+	// TODO: change the role label as `cluster.x-k8s.io/role` sounds very CAPI common (which is not)
+	if azureMachineLabels["cluster.x-k8s.io/role"] == "bastion" {
+		log.Info("DEBUG mario - machine has bastion labels", "machine", machineScope.AzureMachine.Name)
+		// only act on Machines where the NetworkInterfacesReady condition is true
+		azureMachineConditions := machineScope.AzureMachine.GetConditions()
+		for _, condition := range azureMachineConditions {
+			if condition.Type == capz.NetworkInterfaceReadyCondition {
+				log.Info("DEBUG mario - machine has NetworkInterfacesReady condition", "machine", machineScope.AzureMachine.Name)
+				return r.reconcileNormal(ctx, machineScope, clusterScope)
+			}
+		}
+	}
+
+	log.Info("DEBUG mario - machine doesn't met required conditions and labels", "machine", machineScope.AzureMachine.Name)
+
+	return reconcile.Result{}, nil
+}
+
+func (r *AzureMachineReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&capz.AzureMachine{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+func (r *AzureMachineReconciler) reconcileNormal(ctx context.Context, azureMachineScope *capzscope.MachineScope, clusterScope *capzscope.ClusterScope) (reconcile.Result, error) {
+
+	// TODO: implement bastion1 create or update handling
+	// Think about finalizers - maybe we switch to recreate
+	// rolling update means we have to add the finalizer and once the bastion record got deleted, we can remove the finalizer
+
+	log := log.FromContext(ctx)
+	log.Info("Reconciling AzureMachines")
+
+	var bastionIP string
+
+	for _, addr := range azureMachineScope.AzureMachine.Status.Addresses {
+		if addr.Type == v1.NodeInternalIP {
+			bastionIP = addr.Address
+			log.Info("DEBUG MARIO", "azure machine ip", addr.Address)
+		}
+
+	}
+
+	params := scope.DNSScopeParams{
```

Comment:
instead of creating a second `dns-client` per `bastionhost` reconciliation we could also do the following here:

* use the discovered bastion ip and update the already known `cluster/azurecluster` object by adding/updating a new a annotation (e.g. `giantswarm.io/bastionIp`
* let's re-use the value of the `giantswarm.io/bastionIp` annotation in the cluster reconciliation loop [here](https://github.com/giantswarm/dns-operator-azure/pull/70/files#diff-c9f392795aea4d131d38cf4a128c8f6be728ca2ec8b8307cad6ecf1fb7ca5da1R180-R187)
* and do all the further dns-client operations from the cluster reconciliation loop (Instead of having two dns-clients per cluster)